source/hpack/tables.d

```diff
@@ -0,0 +1,340 @@
+		41:	   HTTP2HeaderTableField("if-none-match", ""),
+		42:	   HTTP2HeaderTableField("if-range", ""),
+		43:	   HTTP2HeaderTableField("if-unmodified-since", ""),
+		44:	   HTTP2HeaderTableField("last-modified", ""),
+		45:	   HTTP2HeaderTableField("link", ""),
+		46:	   HTTP2HeaderTableField("location", ""),
+		47:	   HTTP2HeaderTableField("max-forwards", ""),
+		48:	   HTTP2HeaderTableField("proxy-authenticate", ""),
+		49:	   HTTP2HeaderTableField("proxy-authorization", ""),
+		50:	   HTTP2HeaderTableField("range", ""),
+		51:	   HTTP2HeaderTableField("referer", ""),
+		52:	   HTTP2HeaderTableField("refresh", ""),
+		53:	   HTTP2HeaderTableField("retry-after", ""),
+		54:	   HTTP2HeaderTableField("server", ""),
+		55:	   HTTP2HeaderTableField("set-cookie", ""),
+		56:	   HTTP2HeaderTableField("strict-transport-security", ""),
+		57:	   HTTP2HeaderTableField("transfer-encoding", ""),
+		58:	   HTTP2HeaderTableField("user-agent", ""),
+		59:	   HTTP2HeaderTableField("vary", ""),
+		60:	   HTTP2HeaderTableField("via", ""),
+		61:	   HTTP2HeaderTableField("www-authenticate", "")
+	];
+}
+
+private struct DynamicTable {
+	private {
+		// table is a queue, initially empty
+		DList!HTTP2HeaderTableField m_table;
+
+		// as defined in SETTINGS_HEADER_TABLE_SIZE
+		HTTP2SettingValue m_maxsize;
+
+		// current size
+		size_t m_size = 0;
+
+		// last index (table index starts from 1)
+		size_t m_index = 0;
+	}
+
+	this(HTTP2SettingValue ms) @safe
+	{
+		m_maxsize = ms;
+	}
+
+	// number of elements inside dynamic table
+	@property size_t size() @safe { return m_size; }
+
+	@property size_t index() @safe { return m_index; }
+
+	@property ref auto table() @safe { return m_table; }
+
+	HTTP2HeaderTableField opIndex(size_t idx) @safe
+	{
+		foreach(i,f; m_table[].enumerate(1)) {
+			 if(i == idx) return f;
+		}
+		assert(false, "Invalid dynamic table index");
+	}
+
+	// insert at the head
+	void insert(HTTP2HeaderTableField header) @safe
+	{
+		auto nsize = computeESize(header);
+		// ensure that the new entry does not exceed table capacity
+		while(m_size + nsize > m_maxsize) {
+			logDebug("Maximum header table size exceeded");
+			remove();
+		}
+
+		// insert
+		m_table.insertFront(header);
+		m_size += nsize;
+		m_index++;
+	}
+
+	// evict an entry
+	void remove() @safe
+	{
+		assert(!m_table.empty, "Cannot remove element from empty table");
+		m_size -= computeESize(m_table.back);
+		m_table.removeBack();
+		m_index--;
+	}
+
+	/** new size should be lower than the max set one
+	  * after size is successfully changed, an ACK has to be sent
+	  * multiple changes between two header fields are possible
+	  * if multiple changes occour, only the smallest maximum size
+	  * requested has to be acknowledged
+	*/
+	void updateMaxSize(size_t nsize) @safe
+	{
+		assert(false);
+	}
+
+	// compute size of an entry as per RFC
+	private size_t computeESize(HTTP2HeaderTableField f) @safe
+	{
+		return f.name.sizeof + f.value.sizeof + 32;
```

Comment:
`.sizeof` returns the static type size instead of the actual content size. In this case, it looks like this would be more correct:
```
size_t ret = f.name.length + 32;
final switch (f.value.kind) (with HTTP2HeaderTableField.Kind) {
    case str: ret += f.value.get!string.length; break;
    case strarr: ret += f.value.get!(string[]).sum!(s => s.length); break; // + some separator size?
    case status: ret += log10(cast(int)f.value.get!HTTPStatus) + 1; break;
    case method: ret += httpMethodString(f.value.get!HTTPMethod).length; break;
}
return ret;
```
But it seems like it may be necessary to store the entry size along with the value, since AFAIU, the size of the original string representation counts, and that may be ambiguous in some cases (e.g. could a status be reported as "0200"?).